bounter/count_min_sketch.py

```diff
@@ -103,144 +96,85 @@ def __init__(self, size_mb=64, width=None, depth=None, algorithm='conservative')
-            self.logshift = 9
-            self.inc_method = self.log_increment_conservative
-            self.decode = self.log_decode
-
-        self.hash_length = self.width.bit_length()
-        self.hash_mask = (1 << (self.hash_length - 1)) - 1
-        self.logbase = 1 << (self.logshift + 1)
-        self.table = numpy.zeros((self.depth, self.width), self.array_type)
-
-        self.sum = 0
-        self.distinct_counter = HyperLogLog(15)
+        if algorithm and str(algorithm).lower() == 'log8':
+            self.cms = cmsc.CMS_Log8(width=self.width, depth=self.depth)
+        elif algorithm and str(algorithm).lower() == 'log1024':
+            self.cms = cmsc.CMS_Log1024(width=self.width, depth=self.depth)
+        elif algorithm and str(algorithm).lower() == 'conservative':
+            self.cms = cmsc.CMS_Conservative(width=self.width, depth=self.depth)
+        else:
+            raise ValueError("Unsupported algorithm! Use 'conservative', 'log8', or 'log1024'!");
+
+        # optimize calls by directly binding to C implementation
+        self.increment = self.cms.increment
 
     @staticmethod
-    def cell_type(algorithm = 'conservative'):
-        """
-        Returns the numpy dtype used for table cells with provided algorithm
-        """
-        if str(algorithm).lower() == 'logcounter8':
-            return numpy.uint8
-        if str(algorithm).lower() == 'logcounter1024':
-            return numpy.uint16
-        if str(algorithm).lower() == 'logcons1024':
-            return numpy.uint16
-        return numpy.uint32
+    def cell_size(algorithm='conservative'):
+        if str(algorithm).lower() == 'log8':
+            return 1
+        if str(algorithm).lower() == 'log1024':
+            return 2
+        return 4
 
     @staticmethod
     def table_size(width, depth=4, algorithm='conservative'):
         """
         Returns size of Count-min Sketch table with provided parameters in bytes.
-        Does *not* include additional constant overhead used by parameter variables and HLL table, totalling less than 33KB.
+        Does *not* include additional constant overhead used by parameter variables and HLL table, totalling less than 65KB.
         """
-        return width * depth * CountMinSketch.cell_type(algorithm)().itemsize
-
-    def size(self):
-        """
-        Returns current size of the Count-min Sketch table in bytes.
-        Does *not* include additional constant overhead used by parameter variables and HLL table, totalling less than 33KB.
-        """
-        return self.width * self.depth * self.array_type().itemsize
-
-    def bucket(self, i, key):
-        return self.hash_mask & hash_key(i, key)
+        return width * depth * CountMinSketch.cell_size(algorithm)
 
     def increment(self, key):
-        self.inc_method(key)
-        self.distinct_counter.add(str(key))
-        self.sum += 1
+        self.cms.increment(key)
+
+    def __getitem__(self, key):
```

Comment:
```python
def get(self, key, default):
```

might be worth adding, for additional compatibility with the `Counter` interface.

Reply:
That is a method inherited from dict() that doesn't make semantical sense with Counter, which always treats missing values as 0. In fact, CMS is unable to recognize whether a value is missing or 0. 

Using this signature with Counter is either redundant (for default=0) or a bug. Furthermore, we can't even mimic Counter behavior with this method.
I would choose not to support this to prevent errors.

Reply:
If it's in Counter but we don't support it, the proper course of action is a `NotImplementedError` exception.